docs/source/nlp/prompt_learning.rst

```diff
@@ -0,0 +1,378 @@
+
+Dataset Preprocessing
+^^^^^^^^^^^^^^^^^^^^^
+
+The prompt learning dataset accepts a list of json/dictionary objects or a list of json file names where each json file contains a collection of json objects. Each json object must include the field ``taskname`` which is a string identifier for the task the data example corresponds to. They should also include one or more fields corresponding to different sections of the discrete text prompt. The input data might look like:
+
+.. code::
+
+  [
+    {"taskname": "squad", "context": [CONTEXT_PARAGRAPH_TEXT1], "question": [QUESTION_TEXT1], "answer": [ANSWER_TEXT1]},
+    {"taskname": "squad", "context": [CONTEXT_PARAGRAPH_TEXT2], "question": [QUESTION_TEXT2], "answer": [ANSWER_TEXT2]},
+    {"taskname": "intent_and_slot", "utterance": [UTTERANCE_TEXT1], "label": [INTENT_TEXT1][SLOT_TEXT1]},
+    {"taskname": "intent_and_slot", "utterance": [UTTERANCE_TEXT2], "label": [INTENT_TEXT2][SLOT_TEXT2]},
+    {"taskname": "sentiment", "sentence": [SENTENCE_TEXT1], "label": [SENTIMENT_LABEL1]},
+    {"taskname": "sentiment", "sentence": [SENTENCE_TEXT2], "label": [SENTIMENT_LABEL2]},
+  ]
+  
+These additional fields can be unlimited in number and will be used to help map different parts of the discrete text input to a prompt template that you define. We show how this mapping works and how to construct your prompt template in the `_Prompt_Formatting_` section. Data examples for each dataset can all be passed to the dataset class in one file, or in seprate ``.jsonl`` files in a list. 
+  
+.. _data-example-label:
+
+Prompt Formatting
+^^^^^^^^^^^^^^^^^
+
+To customize different prompts for different tasks, we simply need to specify the prompt task template in the config file at ``model.task_templates``. The virtual token markers ``<|VIRTUAL_PROMPT_#|>`` signify where you want virtual tokens to be placed in the template string. ``<|VIRTUAL_PROMPT_0|>``, ``<|VIRTUAL_PROMPT_1|>``, and ``<|VIRTUAL_PROMPT_2|>`` indicate where a number of virtual tokens matching the values given at ``virtual_token_splits[0]``, ``virtual_token_splits[1]`` and ``virtual_token_splits[2]`` will be placed. The other variable fields ``{var}`` refer to the fields in the data json.
+
+For example, given:
+
+- the data json ``{"sentence1": "And he said, Mama, I'm home.", "sentence2": "He didn't say a word."}``
+- virtual token splits set to ``virtual_token_splits = [3, 3, 3]``
+- a prompt template set to ``prompt_template = "<|VIRTUAL_PROMPT_0|> Hypothesis: [sentence1], <|VIRTUAL_PROMPT_1|> Premise: [sentence2] <|VIRTUAL_PROMPT_2|> Answer:"``
+
+the input will be translated into ``VVV Hypothesis: And he said, Mama, I'm home. VVV Premise: He didn't say a word. VVV Answer:``, where ``VVV`` are three virtual tokens.
+
+**We strongly recommend you place all virtual tokens at the very begining of your prompt template** like we do with the ``sentiment`` task example below. We've found this gives the best performance. 
```

Comment:
is this universal that all virtual tokens in front is better than virtual tokens in between?

Reply:
It made a big difference for squad and the authors of the "GPT understands too" paper also make this recommendation, but I should make the recommendation less strong. For other tasks inserting tokens within the prompt might turn out to be better.